### PR TITLE
[13.0][IMP] product_supplierinfo_for_customer: introduce _select_customerinfo

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
This method mimics [standard "brother" `_select_seller`](https://github.com/odoo/odoo/blob/14.0/addons/product/models/product.py#L602) to make the
module more consistent and to introduce a standard way to get the
customerinfo in modules depending on this one.

@ForgeFlow

### CODE REVIEW

**Comment by LoisRForgeFlow** [comment]
cc @pedrobaeza

File: product_supplierinfo_for_customer/models/product_product.py [hunk 1]
```diff
@@ -85,8 +76,8 @@ def price_compute(self, price_type, uom=False, currency=False, company=False):
             partner_id = self.env.context.get(
                 "partner_id", False
             ) or self.env.context.get("partner", False)
-            if partner_id and isinstance(partner_id, models.BaseModel):
-                partner_id = partner_id.id
+            if partner_id and isinstance(partner_id, int):
+                partner_id = self.env["res.partner"].browse(partner_id)
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer/models/product_product.py:80]: Why changing that?

File: product_supplierinfo_for_customer/models/product_product.py [hunk 2]
```diff
@@ -114,3 +105,66 @@ def price_compute(self, price_type, uom=False, currency=False, company=False):
...
+            if customer.product_id and customer.product_id != self:
+                continue
+            if not res or res.name == customer.name:
+                res |= customer
+        # NOTE: This differs from the standard sellers logic. This is done to
+        # keep the old behavior of this module.
+        variant_specific = res.filtered(lambda r: r.product_id)
+        if variant_specific:
+            res = variant_specific
+        return res.sorted("price")[:1]
```
Comment (pedrobaeza) [inline product_supplierinfo_for_customer/models/product_product.py:170]: You are not preparing a hook, but changing more things here, and that should be properly documented and tested. I think we should start with the basic one first.

File: product_supplierinfo_for_customer/models/product_product.py (hunk 1 above)
Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:80]: `_select_customerinfo` like standard `_select_seller` requires a res.partner as argument (later in the funcion it does partner.parent_id), therefore we need to pass the record, not the id

File: product_supplierinfo_for_customer/models/product_product.py (hunk 2 above)
Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:170]: I did this to keep the previous behavior in with a variant specific customerinfo has preference over a non-specific one. This is used to get the price. What do you suggest then? that I isolate this "patch" into a hook?

File: product_supplierinfo_for_customer/models/product_product.py (hunk 1 above)
Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py:80]: Then don't call the variable `partner_id`, but `partner`.

File: product_supplierinfo_for_customer/models/product_product.py (hunk 2 above)
Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py:170]: No, I see the diff, and you are replacing a simple `search` by a lot of code, and that's why I don't get.

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:170]: @pedrobaeza Something like this?
```python
...
        return self._select_customerinfo_hook(res)

    @api.model
    def _select_customerinfo_hook(self, customers):
        # NOTE: This differs from the standard sellers logic. This is done to
        # keep the old behavior of this module.
        variant_specific = customers.filtered(lambda r: r.product_id)
        if variant_specific:
            customers = variant_specific
        return customers.sorted("price")[:1]

```

Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py:170]: Selección_010

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:170]: ah, I see, maybe I should have put a link to the standard "brother": https://github.com/odoo/odoo/blob/14.0/addons/product/models/product.py#L602.

I'm doing the same Odoo does with supplieinfos but for customers, which makes more sense IMO.

Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py:170]: And that's what I say that should go into another PR to be discussed, as this can have side effects, changing the selector for the proper customer info on existing installations.

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:170]: why not discuss it in this one? I pinged you because that was also my concern, if you have installations with this module it can affect you. But I tried to keep the behavior the same, do you see any change that can affect your installations?

Comment (pedrobaeza) [reply product_supplierinfo_for_customer/models/product_product.py:170]: Is there any reason for you to change the current behavior?

Anyways, if changed in this one, some tests for each edge case should be introduced, and compare before and after to be able to see.

About our current customers using this module, @sergio-teruel and @carlosdauden can say better. @chienandalu do you TCR cases?

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:170]: I think adding a method to get customerinfo is almost mandatory, I was reviewing child modules of this one and each of them searches the customerinfo in a different way... 

Once I was doing it, I thought it was good to it good, and for me the best way is to mimic Odoo; I think is an UX improvement if Odoo searches customerinfo and supplierinfo the same way. But this is optional, if you see issues with this I will prepare an alternative.

Comment (LoisRForgeFlow) [reply product_supplierinfo_for_customer/models/product_product.py:170]: Alternative: https://github.com/OCA/product-attribute/pull/891.

**Comment by LoisRForgeFlow** [comment]
This is not going to be done in v13 (hopefully in v14), in the meanwhile it is substituted by https://github.com/OCA/product-attribute/pull/891.
